### PR TITLE
[actions] draft of main.yml

### PR DESCRIPTION
**Pull request by dglaeser**

### CODE REVIEW

**Comment by dglaeser** [comment]
This depends on #21

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      image: continuumio/miniconda3
+
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
```
Comment (dglaeser) [inline .github/workflows/main.yml:25]:
```suggestion
          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
          conda init bash
          source ~/.bashrc
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+          conda activate ~/env_simple_use_case
+
+      - name: run-cwl
+        run: |
+          conda activate $GITHUB_WORKSPACE/env_simple_use_case
```
Comment (dglaeser) [inline .github/workflows/main.yml:30]:
```suggestion
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
+          source ~/.bashrc
```
Comment (dglaeser) [inline .github/workflows/main.yml:27]:
```suggestion
          exec $SHELL
```

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
+          exec $SHELL
```
Comment (dglaeser) [inline .github/workflows/main.yml:27]:
```suggestion
          exec bash
```

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
+          exec bash
+          conda activate ~/env_simple_use_case
+
+      - name: run-cwl
+        run: |
+          cd $GITHUB_WORKSPACE/simple_use_case/cwl
```
Comment (dglaeser) [inline .github/workflows/main.yml:32]:
```suggestion
          exec bash
          conda activate ~/env_simple_use_case
          cd $GITHUB_WORKSPACE/simple_use_case/cwl
```

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
+          exec bash
+          conda activate ~/env_simple_use_case
```
Comment (dglaeser) [inline .github/workflows/main.yml:28]:
```suggestion
```

**Comment by dglaeser** [comment]
@pdiercks, so the last run of the actions went through, although the output of the cwl run is not shown...

**Comment by dglaeser** [comment]
@pdiercks, this depends on #21 to be merged first.

**Comment by dglaeser** [comment]
I'll try to rearrange things in the actions file, but with cf6e7ec503208a1ca711331119e18ddd8cd59105 we have a working version.

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
```
Comment (dglaeser) [inline .github/workflows/main.yml:26]:
```suggestion
          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
          conda init bash
          exec  bash
          conda activate ~/env_simple_use_case
```

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  run-cwl:
+    runs-on: ubuntu-latest
+    container:
+      image: continuumio/miniconda3
+
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
```
Comment (dglaeser) [inline .github/workflows/main.yml:22]:
```suggestion
      - name: prepare-environment
        shell: bash
```

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        run: |
+          cd $GITHUB_WORKSPACE
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
+
+      - name: run-cwl
+        run: |
+          exec bash
+          conda activate ~/env_simple_use_case
```
Comment (dglaeser) [inline .github/workflows/main.yml:31]:
```suggestion
        shell: bash
        run: |
```

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      image: continuumio/miniconda3
+
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        shell: bash
+        run: |
+          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
```
Comment (dglaeser) [inline .github/workflows/main.yml:25]:
```suggestion
          cd $GITHUB_WORKSPACE
          conda env create --file simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
```

Comment (dglaeser) [inline .github/workflows/main.yml:25]:
```suggestion
          conda env create --file $GITHUB_WORKSPACE/simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
```

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        run: |
+          conda env create --file $GITHUB_WORKSPACE/simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+          conda init bash
+          exec  bash
+          conda activate ~/env_simple_use_case
+
+      - name: run-cwl
+        shell: bash
+        run: |
+          cd $GITHUB_WORKSPACE/simple_use_case/cwl
```
Comment (dglaeser) [inline .github/workflows/main.yml:33]:
```suggestion

      - name: run-cwl
        shell: bash
        run: |
          conda init bash
          exec  bash
          conda activate ~/env_simple_use_case
          cd $GITHUB_WORKSPACE/simple_use_case/cwl
```

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+          conda env create --file $GITHUB_WORKSPACE/simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+
+      - name: run-cwl
+        shell: bash
+        run: |
+          conda init bash
+          exec  bash
+          conda activate ~/env_simple_use_case
+          cd $GITHUB_WORKSPACE/simple_use_case/cwl
+          cwltool wf_run_use_case.cwl
```
Comment (dglaeser) [inline .github/workflows/main.yml:34]:
```suggestion
          cwltool wf_run_use_case.cwl
          
          - name: upload-paper-artifact
            uses: actions/upload-artifact@v2
            with:
              name: paper
              path: $GITHUB_WORKSPACE/simple_use_case/cwl/paper.pdf
              retention-days: 1
```

File: .github/workflows/main.yml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+    steps:
+      # Checks-out the repository under $GITHUB_WORKSPACE
+      - uses: actions/checkout@v2
+
+      - name: prepare-environment
+        shell: bash
+        run: |
+          conda env create --file $GITHUB_WORKSPACE/simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+
```
Comment (dglaeser) [inline .github/workflows/main.yml:26]:
```suggestion
```

File: .github/workflows/main.yml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+          conda activate ~/env_simple_use_case
+          cd $GITHUB_WORKSPACE/simple_use_case/cwl
+          cwltool wf_run_use_case.cwl
+
+        - name: upload-paper-artifact
+          uses: actions/upload-artifact@v2
+          with:
+            name: paper
+            path: $GITHUB_WORKSPACE/simple_use_case/cwl/paper.pdf
+            retention-days: 1
```
Comment (dglaeser) [inline .github/workflows/main.yml:41]:
```suggestion
      - name: upload-paper-artifact
        uses: actions/upload-artifact@v2
        with:
          name: paper
          path: $GITHUB_WORKSPACE/simple_use_case/cwl/paper.pdf
          retention-days: 1
```

File: .github/workflows/main.yml [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+          conda env create --file $GITHUB_WORKSPACE/simple_use_case/cwl/default_env.yml --prefix ~/env_simple_use_case
+
+      - name: run-cwl
+        shell: bash
+        run: |
+          conda init bash
+          exec  bash
+          conda activate ~/env_simple_use_case
+          cd $GITHUB_WORKSPACE/simple_use_case/cwl
+          cwltool wf_run_use_case.cwl
```
Comment (dglaeser) [inline .github/workflows/main.yml:34]:
```suggestion
          cwltool wf_run_use_case.cwl
          ls
```

**Comment by pdiercks** [comment]
okay, thanks for figuring this out! it seems to be working, but is there anything you would want to add? As I am not really at home with github actions, were is the final paper.pdf uploaded to?
I will merge #21 first and then this can be merged too?

**Comment by dglaeser** [comment]
@pdiercks, I would say #21 can be merged but I would wait with this one. I am not sure if it is working properly, because the artifact upload fails. There are a lot of fixup commits here which should be squashed as well. I have no permission with my access token to change the `main.yml` file, but it works when I edit it directly here online. I probably have to generate some special token or so, haven't yet checked that. But that's why I cannot yet clean the history, and moreover,  there is some debug output still in the cwl job. But for some reason we don't see the output of the cwl run.. I suspect it is actually not yet working although the final status is set to "passed"...

**Comment by dglaeser** [comment]
Besides this, once I figure out if this works, or how to make it work, I would also add the execution of the other workflows to the pipeline within this PR. What do you think?

**Comment by pdiercks** [comment]
Okay, let me know if I can be of help. 
Regarding the other workflows: at the moment e.g. nextflow and snakemake handle the software deployment and they would create conda environments on their own, such that for running that in the pipeline one would install nextflow (snakemake respectively) and then execute the command to run the workflow.
Or would you rather use one environment for all workflows?

**Comment by dglaeser** [comment]
I would define the other workflows as other jobs, so they would run in parallel and the environment setup in the cwl job does not leak into the others. I would say we let the tools handle the environment stuff, no?

**Comment by pdiercks** [comment]
okay, sounds good to me.

**Pull request closed without merging**